---
title: "Working with diatom data in the `neotoma2` R package"
output:
  rmarkdown::html_vignette: default
  html_vignette:
    toc: yes
    toc_depth: 4
vignette: |
%\VignetteIndexEntry{Working with diatom data in the `neotoma2` R package} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup}
library(neotoma2)
library(dplyr)
library(purrr)
library(sf)
```

# Start from line 710 March 31th
## Introduction

This document provides general guidance and specific code examples for those wanting to use `neotoma2` to retrieve and work with diatom data from the Neotoma Paleoecology Database. This vignette, as part of the `neotoma2` package, also provides the package developers (Socorro Dominguez and Simon Goring) a set of structured descriptions for the kinds of work that diatom researchers may wish to perform with Neotoma data and some of the specific tasks they may wish to accomplish. Researchers who would like to add examples or other workflows can reach out to the developers through [GitHub Issues](https://github.com/NeotomaDB/neotoma2/issues), or the [Neotoma Slack Workspace](https://join.slack.com/t/neotomadb/shared_invite/zt-cvsv53ep-wjGeCTkq7IhP6eUNA9NxYQ).

Some code in this document was borrowed from the `neotoma` markdown document prepared by Simon Goring and the original Neotoma R package ([Goring et al. 2015](http://doi.org/10.5334/oq.ab)).

One of the main goals of `neotoma2` is to pull records out of the Neotoma database and into a user's R workflow. This will help diatomists and environmental scientists produce clear working documents that use data from Neotoma and allow them to combine it with their own data or data from other sources.

> If you're thinking without writing, you only think you're thinking. — *Leslie Lamport*
This document is organized to address users' two main goals: to browse the database to discover datasets that meet specified criteria and to create output files that can be used for further analysis. Other goals are to create plots and maps and analyze data using R code.

### Why `neotoma2` and not just an update of `neotoma`?

The Neotoma servers and all the back-end services recently (as of [Simon - add month / year]) migrated from a SQL Server Database on a lonely Windows N95 computer to a more robust server and set of services. Because of this migration, many of the tools we use to access data are no longer accessible. For this reason, the Neotoma IT Team chose to create a new package. This is because the code used in `neotoma` will not be directly comparable to the code used in `neotoma2` since data objects and expected behaviours will differ.

### Data Organization in Neotoma
[[Re following two sections.  Can they be combined? - Delete the What does the data loook like heading?  Can SQL be used to work with the Postgres database?]]
It is worth knowing how the Neotoma database is organized and what we mean when talking about things like `sites`, `datasets`, and `samples`. The database is a Postgres database, and people who are comfortable with SQL can download a snapshot of the database and run much of the analysis on their own.

#### What does the data look like?

The Neotoma data model is available online at the [Database Schema](https://open.neotomadb.org/dbschema) page. As of September 2021, there were 150 different tables in the database. Some of these have particular kinds of data (for example, `agetypes`), and some link tables together. The `neotoma2` package directly supports access to some tables, but not all. For specialized cases, a user may be required to access an entire table directly (using the `get_table()` function).

#### Sites, Datasets, etcetera

Neotoma data is structured spatially. Each collection location is assigned a **site**, a spatial location defined by either a point or a polygon. It may be a lake, an archaeological site, a swamp, or another location.

Within a **site**, there may be one or more **collection units**.  
A collection unit represents the material (e.g., sediment core) removed from a specific location within a `site` (e.g., a lake). So, a researcher may have taken multiple cores within a lake basin, or an archaeologist may have several dig sites within a larger area. Collection unit material is subdivided into **analysis units** (e.g., sediment intervals within a core). **Samples** can be removed from analysis units and analyzed for a particular dataset type (e.g., diatoms, pollen, water chemistry). Groups of analysis units (samples) from an individual collection unit are then grouped into a **dataset**.

#### What is the R package doing?

The `neotoma2` package uses data from the Neotoma database, but it is mediated through R functions (like `get_sites()` to get site information) that call public URLs, which are abstractions of database calls. So, for example, to get data for the site with ID number 12:
```r
get_sites(12)
```

calls out to the internet for the result at:
```
https://api.neotomadb.org/v2.0/data/sites/12
```

Which itself calls an SQL query:
```sql
SELECT * FROM ndb.sites
WHERE siteid = 12
```

It is a lot easier for most users to use the R functions since they create objects in R that work with other commands and statistical analysis they might want to do.

#### Functions in `neotoma2`
Following are descriptions of the main functions in `neotoma2`. Text in quotes was taken from the package descriptions. See the package documentation for more details.

`neotoma2` was being developed while this vignette documentation was being prepared, so not all the package functions were available. Currently (March 2022), some functions from `neotoma` and `neotoma2` are used in this document. Over time, original package functions should be replaced with corresponding `neotoma2` functions. In the code chunks below, `neotoma::` refers to functions in the original neotoma R package, for example, **`neotoma::`**`get_site(whatever)`.

##### `neotoma2` functions

There are **3** main functions in `neotoma2` to retrieve data: `get_sites()`, `get_datasets()`, and `get_downloads()`


## Browsing - Data discovery

This section describes ways of searching the Neotoma database to find `sites` and diatom `datasets` using various criteria. The primary purpose of each code chunk in this section is to find the `sites` or `datasets` that meet the criteria specified in the code. The returned objects can then be used as input of other code chunks that return more detailed information on the datasets and create output files for data analysis (e.g., diatom counts).

###### get_sites
**Usage:** Sites can be searched using the following supported parameters.
e.g., `get_sites(sitename="Alexander Lake")`

| parameter | allowed values |
| ------ | ------- |
| `siteid` | integers (values > 0) |
| `sitename` | valid character string; `%` as wildcard |
| `altmax` | integers (values >= than `altmin`) |
| `altmin` | integers |
| `loc` | spatial coordinates (WKT or geoJSON) |
| `limit` | integer > 0; default 25 |
| `offset` | integer > 0; default 0 |

Values returned: A sites object that includes: siteid, sitename, long (mean), lat (mean), elev, description, long_acc [width of bounding box], and lat_acc [height of bounding box]

`limit` indicates the maximum number of records that will be returned; `offset` indicates the number in a sequence where processing will begin.  Together, these allow large numbers of records to be processed in batches.

To see structure of `get_sites()` function output, run the following code:

```{R structure_of_get_site_output}
uwallf_site <- get_sites(sitename = 'Upper Wallface%') # finds site data for a site
#str(uwallf_site) # shows structure of uwallf_site
```
### Find sites

We can search for information on a single or multiple sites with `neotoma2::get_sites()`. Sometimes it is helpful to find locations of all sites within a region, even if they are not diatom sites, as this indicates the level of paleoecological 'effort' in a region. We can often use a site search to do preliminary work since the `sites` object contains less information than the `datasets` or `downloads` objects. 
For example, we might want to generate a map of all sites and superimpose specific sites we are interested in.

#### Site Name

If we know the names of one or more of the existing sites we are looking for, we can get site data using the `sitename` parameter.

Get data on a single site using sitename

```{r sitesbyName example}
# diat <- get_sites(sitename = "")  # Basic format
diat <- get_sites(sitename = 'Schrader Pond')  # Example
diat
```


```{r sitesbyNameWildcard example}
diat <- get_sites(sitename = '%Schra%')      # Example using wildcards (%)
diat
```

Get data on multiple sites using sitenames.

```{r multsitesbySitename example}
# Join a list of site names together using the combine command, c()
sitenameList <- c("Bald Hill", "Blake (sutton)", 
                  "High (sudbry)", "Holland", 
                  "Little Hosmer", "Long (grnsbo)", 
                  "Long (shefld)", "Round (shefld)")
my_sites <- c()
for(sitename in sitenameList){
  a_site <- get_sites(sitename = sitename)
  my_sites <- c(my_sites, a_site)
}
my_sites
```

Get data on multiple sites using sitenames - National Lakes Assessment Example

#### Site ID
Find sites using siteid.  In the following code chunks, we assume that siteids have been obtained from Neotoma Explorer, from Neotoma Landing Pages or from an external file.

Get data on a single site using siteid
```{r sitesbySiteID example}
# diat_3 <- get_sites(x) # basic format
diat_4 <- get_sites(27408) # Example for single site (Schrader Pond)
diat_4
```

Get data on multiple sites by using siteids.
| `sitename` |  `siteid` |
| -- | -- |
| Jackson Pond | 25918 |
| Jellison Hill Pond | 25919 |
| Jimmie Pond | 25920 |

Get data on a multiple sites by using siteids
```{r multsitesbySiteIDs example}
siteids <- c(25918, 25919, 25920)
diat_6 <- get_sites(siteids)
diat_6
```
Get data on multiple sites by reading a text file of siteids

The above code returns data for three sites.  We could use a larger set of site IDs either by loading the IDs from a file, or by entering them directly.  Here we assume that there exists a file called `NLAD.txt` with a list of site IDs in a plain text file for the National Lakes Assessment.  The contents of the file look like this (the ellipsis would be replaced by other numbers):


There may well be many more lines with numbers representing each site identifer in the set of records.  In this case we would load the data like so.:

```{r NLAmultiplesiteID}
nlad <- read.table('NLAD.txt', header=FALSE)
nlaSites <- get_sites(nlad$V1)
nlaSites
```

#### Geographic location
Find all sites located in a geopolitical unit as specified by a gpid.

View the geopolitical table to search for locations and gpids.

A gpid is a useful way to specify country and state boundaries.
The code below will open the Geopolitical units table.  Once the table appears, use the search function at top of the page to find the name of a geopolitical unit of interest and its corresponding gpid. 

View the site geopolitical table to find gpids associated with datasetids

```{R geopoliticalunitsTable}
loc <- neotoma::get_table(table.name ='geopoliticalunits')
View(loc)
# Use the filter at the top of the page to find name of a geopolitical Unit and its corresponding gpid
```

**Still not possible to do but this is how it should work**
```{r get site by gpid}
# diat_1<-get_site(gpid = x) # basic format
diat_1<-get_sites(gpid = 6198) # Example using geopolitical ID for Alaska - GeoPoliticalUnits table 
diat_1
```


Run following code to open the sitegeopolitical table and then enter datasetid in filter box to find corresponding gpids

**Not sure what this table is** <- Socorro
```{R sitegeopoliticalTable}
dataset_gpid <- get_table(table.name ='sitegeopolitical')
View(dataset_gpid)
# Table columns: GeoPoliticalID,  SiteID, SiteGeoPoliticalID
```

Example - get sites from multiple geopolitical units (e.g., states)
**NOTE: gpid not returning results**
```{r retrieveByGeopol}
# Get site data for all sites in three New England states (ME, VT, NH, CT, MA, RI) plus NJ and NY [DFC - delete some states - only 3]
# The gpids were obtained from the Geopolitical units table 

gpids <- c(8412, 7990, 7934)

ne_sites <- c()
for (id in gpids) {
  a_site <- get_sites(gpid = id)
  ne_sites <- c(ne_sites, a_site)
}

ne_sites
```

#### Elevation
Find sites between minimum and maximum elevations (m) 

```{R sitesbyElev}
diat_el <- get_sites(altmin=100, altmax=120)
```

###### get_datasets
**Usage:** get_datasets(x, datasettype, piid, altmin, altmax, loc, gpid, taxonids,taxonname, ageold, ageyoung, ageof, subdate)

x is an optional value that represents the numeric dataset ID.

`get_datasets()` returns a **sites** object but with a more complex detail at the dataset level.

The extra values returned for each dataset are: database, age_range_old, age_range_young, notes, location, waterdepth, collunitype, collectiondevice, collectionunitname, collunithandle.

To see structure of a dataset after `get_datasets()`, run the following code:
```{R structure_of_get_dataset_output}
uwallf_site <-get_sites(sitename = 'Upper Wallface%') # finds site data for a site
uwallf_datasets <- get_datasets(uwallf_site) # finds datasets data for Upper Wallface (there's 4 elements)
uwallf_diatom_dataset <-  uwallf_datasets[[1]]@collunits[[3]]@datasets[[2]] #(datasetid is 49249)
#str(uwallf_diatom_dataset ) # shows structure of uwallf_datast
```

#### Dataset ID
Get meta data on a single dataset using datasetID 

```{r get datasetID}
# diat_ID <- get_datasets(datasetid) - basic format
diat_ID <- get_datasets(32252) # Example
diat_ID
```

```{r TODO Create browse function}
#browse (44764) [Simon - opens Neotoma Explorer, but does not go to the site specified]
```

Get data on multiple datasets using multiple dataset IDs 
```{r multdatsetid }
multid <- c(44764,44765,44766)
diat_multid <- get_datasets(multid) # Example
diat_multid
```

```{r GetSampleIDs by Dataset IDs}
NL1 <- c(49651,49032,49033,49034,49035,49036,49037)
NL2 <- get_datasets(NL1)
datasets(NL2)
```

#### Dataset Type
`neotoma2` can retrieve datasets from the API based on their dataset type (Diatom, Diatom Surface Sample, Diatom Top-Bottom)
Example below gets dataset information for individual dataset types and then combines results into one object.

Usually, code would include selection criteria in addition to datasettype. For example, all datasets of a particular dataset type within a specific geographic region. 

[old text - should be deleted: Example "loc = bounding box" (West,South,East,North) is northern Maine.]
```{r getDiatomsByDatasettype} 
# Get datasets by individual diatom datatype 
neDiatoms_ss <- get_datasets(datasettype="diatom surface sample", limit=30)
neDiatoms_ss 
```


```{r}
neDiatoms_tb <- get_datasets(datasettype="diatom top-bottom")
summary(neDiatoms_tb)
```
```{r}
diatom_list <- c("diatom", "diatom top-bottom", "diatom surface sample")
neDiatoms_com <- c()
for (type in diatom_list){
  sites <- get_datasets(datasettype=type)
  neDiatoms_com <- c(neDiatoms_com, sites)
}
neDiatoms_com
```

#### Geographic region

#####  Geopolitical units (e.g., states)

`get_datasets` for an individual state - Pennsylvania examples **NOTE code not ready for gpids**
```{R gpidExamplePenn}
# Get_dataset by name
Geo_PA <- get_datasets(gpid = "Pennsylvania")
# Get by gpid
GEO_PA1 <- get_datasets(gpid = 8412) # Pennsylvania gpid = 8412
```

##### Location based extraction

###### Latitude and Longitude bounding boxes

```{R diatomsbyBBox}
# diat_LL <- get_datasets(datasettype="diatom", loc=BOUNDING BOX) - basic format
```

Example of get_datasets by location with bounding box of lat & longs (West, South, East, North)

```{R diatomsbyBBoxExecuted}
diat_NE <- get_datasets(datasettype="diatom",loc =c(-80,40,-65,48)) 
plotLeaflet(diat_NE)
```


###### User-specified spatial polygon

With the `sf` package to create polygons for spatial searches (the package also allows you to pass in geoJSON or WKT spatial objects). We can pass a polygon as a in the `loc` argument of `get_datasets()` in `neotoma2`. 

In the example below, we first create a matrix to represent the corners of a bounding box of the northeastern United States. We then create a spatial polygon ("neus") based on the co-ordinates.  Finally, we to use "neus" as input to the `get_datasets()` function.

```{R siteSpatialSearch}
northeastUS <- rbind(c(-80,40), c(-80,48),
                     c(-65,48), c(-65,40), c(-80,40))
neus <- sf::st_polygon(list(northeastUS))
``` 

```{r}
neUSsites <- get_datasets(loc = neus)
neUSsites
```


#### Elevation (m)
Find datasets for sites between minimum and maximum elevations

```{R getSitesByElevExample}
diat_el <- get_sites(altmin=500,altmax=5500)
diat_el
```
```{r}
diat_el2 <- get_datasets(diat_el)
diat_el2
```

###### get_downloads
Using the dataset ID, site object or dataset object, return all records associated with the data as a `sites` object with an even more complex detail level.

**Usage:**  `get_downloads(x)`
Arguments: `x` can be a dataset ID or a vector of numeric dataset IDs. It can also be a `sites` object as returned by `get_datasets()` or `get_sites()`; verbose, logical printing a download status bar.

The values are returned as a `sites` object with a full level of detail; relevant assemblage information and metadata.

The downloaded result is a sites object that contains the constituent taxa, the chronology, site and PIs who contributed the data. 

The following are important components: datasets, chronologies, samples within the dataset.

To see structure of `get_downloads()` function output, run the following code:
```{R structure_of_get_download_output}
uwallf_site <-get_sites(sitename = 'Upper Wallface%') # finds site data for a site
uwallf_datasets <- get_datasets(uwallf_site) # finds datasets data for Upper Wallface (there's 4 elements)
uwallf_diatom_dataset <- neotoma2::filter(uwallf_datasets, datasettype == "diatom") # only keep diatom datasets
uwallf_download <- get_downloads(uwallf_diatom_dataset)
#str (uwallf_download) # shows structure of uwallf_datast
```


Get Sample IDs for Dataset IDs
```{r GETSampleID for Dataset ID}
CODE <-c(44764,44765,44766) #datasetids
C1<- get_datasets(CODE)
CD <- get_downloads(C1)
C2 <- samples(CD)$taxonid
C2
```

```{r}
CD
C3 <- datasets(CD)
as.data.frame(C3)$datasetid
```
```{r}
C4 <-collunits(CD)
C4
```

```{r GetSampleIDs by Dataset IDs and filtering by type}
NL1 <- c(49651,49032,49033,49034,49035,49036,49037)
NL2 <- get_datasets(NL1) %>%
  neotoma2::filter(datasettype == "diatom surface sample")
NL3 <- get_downloads(NL2)
NL4 <- samples(NL3)$taxonid
NL4
```

## Using neotoma2::filter

There might be cases when we need to filter results out from a `sites` object. We can use the `filter()` function. `filter()` takes any `sites` attribute to filter on.


Find sites where value for altitude is null

```{r Find sites where elev is null}
diat_el <- get_sites(limit = 1000) %>%
  neotoma2::filter(is.na(altitude))
diat_el
```

#### Within or intersecting a specified time interval 

```{R Diatom Time Interval}
NL1 <- c(49651,49032,49033,49034,49035,49036,49037)
NL2 <- get_datasets(NL1)

datasets(NL2)

interval_42 <- NL2 %>%
  neotoma2::filter(age_range_old > -42)
interval_42

get_downloads(interval_42)
```

#### Investigator name / person name
[Following code is for how to find a contact id given a contact name; still need code for how to find all datasets associated with a contact]
```{R Diatom Investigator}
Don <- get_contacts(contactname = '%Charles, Donald F.%') 
# In addition to contactname, other possible arguments are: contactid, contactstatus, are familyname
Don
```

#### Taxon name or ID 
** Note taxon name / id not sure if in API)
```{R getTaxa}
# Find datasets using taxon name (doesn't work)
diat_nm <- get_datasets(taxonname = 'Navicula minima')
# Find datasets using taxon id (doesn't work)
diat_nm1<- get_datasets(datasettype = 'diatom',taxonids = 11249)
# Find datasets of a specific data type using taxon name
# [Simon - need code for finding datasets for more than one taxon at a time]


#To access the diatom taxa table use get_table('Taxa'), Taxon name and Taxon ID can be used to find taxa by dataset types
diat_taxa <- get_table('Taxa')
# view (diat_taxa) # run this line to open searchable copy of Taxa table
```

```{r datasettype and filter taxons}
# Find datasets of a specific data type
diat_nm1<- get_datasets(datasettype = 'diatom')

# Get the downloads detail to retrieve the taxons. Apply samples() to the downloads object
diat_nm1_dw <- get_downloads(diat_nm1)
diat_samples <- samples(diat_nm1_dw)
diat_samples

# Use dplyr::filter to obtain the taxa of interest
diat_samples <- diat_samples %>%
  dplyr::filter(variablename == 'Navicula minima')
```

#### Filtering by Submission date

```{r SubDate}
# Filtering by submission date, YYYY-MM-DD
as.data.frame(collunits(diat_nm1_dw)) %>%
  dplyr::filter(colldate == "2007-06-05")
```

```{r filtered collunits with site and dataset id}

dw_df <- as.data.frame(collunits(diat_nm1_dw)) %>%
  dplyr::filter(colldate == "2007-06-05")

ids_df <- getids(diat_nm1_dw)

left_join(x=dw_df,y=ids_df, by=c("collectionunitid" = "collunitid"),all.x=TRUE) %>%
  select(siteid, datasetid, collectionunitid, handle, colldate)

```

#### Dataset DOI

** Not sure dois is a used parameter.
```{r getDiatomsByDOI}
# neDiatoms <- get_datasets(dois=c(...)) # basic format
neDiatoms <- get_datasets(dois = c(43516, 43519, 46228)) # Example using DOI's for Livingston P, West P and Wolf L in the Adirondack Mountains, NY.
# [SIMON - This code chunk does not work.  I assume this is because the Neotoma2 get_datasets function is still being developed. True?  My understanding is that it would retreive the DOI. Don]
```

#### Collection unit type, ID or Handle [secondary]
Code to retreive datasets by collection type - usually would be one part of a larger code chunk or function

```{R diatombyColltype}
# Note: Couldn't find collectionunits of type core. It would be the same code but replacing
# filter(collectiondevice to collectiontype)

df1 <- as.data.frame(collunits(diat_nm1_dw)) %>%
  filter(collectiondevice == "Modified KB gravity corer")

id_df <- getids(diat_nm1_dw)

left_join(df1, id_df, by = c("collectionunitid" = "collunitid"))$datasetid
```

```{R sitesByCollUnitHandle}
df1 <- as.data.frame(collunits(diat_nm1_dw)) %>%
  filter(handle == "ROCKY15")

id_df <- getids(diat_nm1_dw)

left_join(df1, id_df, by = c("collectionunitid" = "collunitid")) %>%
  select(siteid, datasetid, collectionunitid, handle, location)

dataset_rocky15 <- get_datasets(18047) #18047 is the datasetid obtained from above
dataset_rocky15
```

### Find datasets (multiple functions)
#### Constituent database
[SIMON - Code needs to be added]
```{R SitesbyConDB}
consDB <- get_
```

Example of finding constituent databases

```{R SitesbyConDBExample}
ConDB <- 
  
  ```

#### Sample keyword [secondary]
Code to retreive datasets by keyword - usually would be one part of a larger code chunk or function
[Code needs to be written]

Example of using the search function returning keywords 
```{R KW in notes}
sediment_txt <- as.data.frame(datasets(diat_nm1_dw)) %>%
  dplyr::filter(grepl('Sediment sample', notes))

sediment_txt$datasetid

```

#### Aggregate dataset(s) <- ??
Get dataset IDs of all members of an aggregate dataset
Also, get_datasets and get-downloads
```{R Diatom Aggregate Dataset}
aggr_set <- read.csv()# If we are going to use the Excel file that contains sampleids for each aggregate dataset, we should read it in here
aggr1 <- get_datasets() # Do we need to use get_datasets if we already know sampleids?
aggr2 <- get_downloads()
# Code needs to be written
```

Example
```{R Diatom Aggregate Dataset Example}
```

#### Publications
Get all datasetids associated with a particular publication

```{R Diatom Publication}
diat_pub <- get_datasets()
```
Example - [not sure what this code does...] What is the goal?? Socorro
```{R Diatom Publication example}
diat <- get_datasets(datasettype = "diatom")
diat_pub <- get_publication()
pub_diat <- sapply(diat_pub, 
                   function(x) {
                     x$pi.data$ContactID
                   })
```

#### Water chemistry
##### Is water chemistry available?
Determine if any water chemistry data is available for one or more sites or diatom datasets for those sites
If chem data are available, which parameters? (List variables included)
[SIMON - Code chunks in this section do not work]

##### From sites within range of water chemistry characteristics
Get datasets associated with water chemistry within a specific range (e.g., pH between 6 and 8)

Example of how to get water chemistry from the different dataset types [includes snippets of code, none of which work]

```{r getWaterChemEX}
diatomss <- get_datasets(datasettype = 'diatom surface sample') %>% 
  get_downloads()

# Following 3 lines of code seem to work, but search through ALL water chemistry datasets
waterchem <- get_datasets(datasettype = 'water chemistry') %>% 
  get_downloads() 
```

**Note Socorro: not clear what is supposed to happen here; dont find any property called pH so tried with waterdepth.
```{r}
as.data.frame(collunits(waterchem)) %>% 
  dplyr::filter(waterdepth < 5)

# Below - for the future....  get_aggregate function does not yet exist
diatomss <- get_aggregate(123)
waterss <- get_aggregate(124)
diatomdl <- get_downloads(diatomss) %>% compile_downloads()
waterdl <- get_downloads(waterss) %>% compile_downloads()
# Following proposed function would match corresponding diatom and water chemistry samples
aligned <- align_sites(diatomss, waterss)
```

## Need explanations on here (?)
```{r fixing}
diatomss<- get_dataset(datasettype = 'diatom surface sample')
waterss<- get_dataset(datasettype = 'water chemistry')
diatom<- get_dataset(datasettype = 'diatom')
#Sapplyy function is used to return diatomss, read / extract the metadata and select dataset IDs for the first 20 lakes 
subsetss <- sapply(diatomss, 
                   function(x) {
                     x$dataset.meta$dataset.id
                   })[1:20]
#Sapplyy function is used to return diatom, read / extract the metadata and select dataset IDs for the first 20 lakes
# [Simon - I am not sure what the following code does - please document] ?
subset <- sapply(diatom, 
                 function(x) {
                   x$dataset.meta$dataset.id
                 })[1:20]
if('filename' %in% list.files()) {
  diat_dl <- readRDS('filename') # It returns the file that is used 
} else {
  diat_dl <- get_download(subset)
  diat_dlss <- get_download(subsetss)
  diat_dl <- bind(diat_dl, diat_dlss)
  saveRDS(diat_dl, 'filename') # Saves the download object as a file
  #  Where is "filename"saved? - What directory? - type getwd() in the console to find out where the file is.
}
```



##### From sites where one or more specified water chemistry characteristics were measured
```{R Diatom Multiple Water Chemistry}
library('WriteXLS')
diatomwch<- get_datasets(datasettype = 'water chemistry')
diatom<- get_datasets(datasettype = 'diatom surface sample')

subsetwch <- getids(diatomwch)$datasetid[1:20]

subset <- getids(diatom)$datasetid[1:20]

if('fileWCH' %in% list.files()) {
  diat_dl <- readRDS('fileWCH') # It returns the file that is used 
} else {
  diat_wch <- get_downloads(subsetwch)
  diat_dls <- get_downloads(subset)
  diat_dlb <- c(diat_wch, diat_dls)
  saveRDS(diat_dlb, 'fileWCH')
}
# fileWCH will be created in the working directory
#wideWCH_DTM <- compile_downloads(diat_dlb)  # what is this supposed to do?
## neotoma function converts multiple downloads into single object 
minWCH <- as.data.frame(collunits(diat_dlb))

m_ids <- getids(diat_dlb)

minWCH <- left_join(minWCH, m_ids, by = c("collectionunitid" = "collunitid")) %>%
  group_by(datasetid) %>% #dataset was returned from the table 
  summarise(min(waterdepth)) %>% #dplyr function to create a dataframe
  rename(waterdepth = "min(waterdepth)")

wideWCH_DTM <- minWCH %>% 
  filter(is.na(waterdepth))
wideWCH_DTM
# Write data to an Excel file, which will be created in the working directory
WriteXLS(wideWCH_DTM, "wideWCH_DTM.xlsx")
```
## Socorro from here April 4th

#### Diatom Sample analyst name
Get all datasets associated with name of a sample analyst
[SIMON - Currently seems to just get a list of people who analyzed surface samples; also, may not get diatom analysts - gets investigators associated with all "diatom" datasets instead; will need to involve get_downloads]
Needs to be able to accept contact ID of a specific person
```{R Diatom Sample Analyst}
Con_<- get_dataset
Contact_diat <- get_dataset(datasettype = 'diatom')
Analyst <- sapply(Contact_diat, 
                  function(x) {
                    x$pi.data
                  })
```
Example
[Simon - Can this code be modified so it retrieves all dataset IDs associated with an specific example person?]
```{R Diatom Sample Analyst example}
Con_<- get_dataset(32252)
Contact_diat <- get_dataset(datasettype = 'diatom')
Con_U <- sapply(Con_, 
                function(x) {
                  x$pi.data
                })
```


#### Diatom taxa
##### Diatom counts that include one or more specific diatom taxa
[SIMON - Code does not include possibility of specifying a particular diatom taxon]

Get list of taxa found in a single dataset (core)

```{R datasetByTaxa}
diat_tax <- get_datasets(32252) # Datasetid for U Wallface P core 1
diat_dl <- get_downloads(diat_tax)
diat_taxa <- taxa(diat_dl)
diat_taxa
```

Get list of all taxa associated with single datasettype 

```{R datasetByTaxaExample}
diatoms<- get_dataset(datasettype = 'diatom')
diat_dl <- get_download(diatoms)
diat_taxa <- taxa(diat_dl) # List of taxa in all datasets of dataset type "diatom"
# Takes a long time to run
```


##### Diatom counts include only taxa with abundance greater than x percent (in sample or core)

```{R Diatom Counts}
```

Example

```{R Diatom Counts EX}
```

#### Site characteristics
##### Site area larger or smaller than specified size

```{R Diatom Site Area}
```

Example

```{R Diatom Site Area EX}
```

#### Depository in which diatom slides are located

```{R Depository}
diat.data <- get_dataset(datasettype='diatom') # access to neotoma explorer but needs to specify 
browse(diat.data)  #This opens Neotoma Explorer, but does not show any data
```

Example
```{R Depository EX}
library(neotoma)
```
#### Chronology (y/n) and characteristics
Does a dataset have an associated chronology?  If so, what are metadata (e.g., upper and lower age bounds)?
Not sure of purpose of the following code
```{R Chron }
get_datasets(32252) %>% # datasetid for Upper Wallface Pond
  get_downloads() %>%
  chronologies() 
# (old text) related to how a bacon model is run through R
```
Example
```{R AnotherChronExample}
```

Example
```{R ChronologyExample}
```


### Format for browsing using multiple criteria - Examples 
Examples: [could get complicated]  General rules and tips.  [Add as they become known]
### [Following sections to be reorganized]
#### Find all diatom datasets of any datatype

#### Find all diatom datasets of any datatype within a geographic region

#### Find all diatom surface sample datasets within a geographic region

#### Find datasets .... with surface area larger than x ha .......

### Other options / Misc:

### Create output of browse results 
Minimum “list” of dataset, collection, and / analysis unit ids for further browsing and use as input to other R functions. 

Standard comprehensive output file for use in detailed browsing. Variables should include all of the following (variables are grouped by their location in Tilia tabs): 
1.  Site information: Site ID, Site name, Latitude, Longitude, Latitude width, Longitude width, Country, State, County, Admin Unit (3rd Geographic Unit), Lake characteristics, publication id
2. Collection unit information:  CU ID, Type, Collection unit name, Collection unit handle, Location in site, Collection device, Date collected, Water depth of collection
3.  Dataset information: Data Type (Diatom, Diatom surface sample, or Diatom top-bottom), Primary publication, Investigators [or first author of publication], Sample analyst


## Create output files for data analysis

Code chunks in this section are designed to create output files for diatom counts, metadata, chronology, water chemistry, and other kinds of data.  They provide examples for retrieving data and creating output files in a few basic formats.  Formats include those that can be used as input to data analysis programs - R packages and others.  In general, though they do not do so now, basic output files should include all related data; the number of variables can be reduced later.They will include several descriptive variables and be easily readable. [input to code chunks can include output from browse tasks]

### Basic files / matrices

#### Diatom counts
The following code chunks can be used to view diatom count related data.  There are a few examples, each with something a little different.
For single sites and multiple sites; aggregate data sets; stratigraphic vs surface sample vs top-bottom; percentages vs raw counts; containing specific taxa, etc.

##### View sample and count metadata
```{R Upper Wallface View counts}
### View sample metadata, list of taxa, and counts in Download dataframes 
### View sample metadata,
uwallface.dw[["32252"]][["sample.meta"]] 
### View taxa list
uwallface.dw[["32252"]][["taxon.list"]]
### View diatom counts
uwallface.dw[["32252"]][["counts"]]
### View Chronology
uwallface.dw[["32252"]][["chronologies"]]
```
##### Count data for one or more datasetids
Gets count data for one or more datasets and writes data to an Excel file.
```{R ExampleDataset}
library(purrr)
V1 <-c(44786,44787,44788) # Datasetids
# In this example we combine datasets from 3 sites (New England lakes).  To get output for aggregate datasets, a file with all the datasetids for an aggregate dataset could be used as input.
V1_diatom <- map(V1, function(x)
  neotoma:::get_dataset(x,datasettype="diatom")) %>%
  neotoma::bind()   
#In the get_dataset line, whatever is in the parenthesis (a parameter such as x or gpid); will be whatever is in the object V1, defined in the line above.
# [SIMON - Why does datasettype need to be specified above, given that input data are datasetids?  The code seems to run fine with "datasettype="diatom" removed.]
if('V1' %in% list.files()) {
  V1_dl <- readRDS('V1') 
} else {
  V1Datasets <- list()
  #
  for(i in 1:length(V1_diatom)) {
    V1Datasets[[i]] <- get_download(V1_diatom[[i]])
  }
  #For loop will go through each step to first create the file and then download data for each site
  V1_download <- neotoma::bind(V1Datasets)
  saveRDS(V1_download, 'V1')
}
V1Counts <- compile_downloads(V1_dl)
V1Row <- V1Counts %>% 
  group_by(dataset) %>% 
  summarise(min(depth))
WriteXLS(V1Counts,  "VL_dl.xlsx")
#This last step creates a table in an Excel file which will be saved in the working directory
```

##### Count data file for a single site using site name
Following code uses Neotoma2 R - Schrader Pond example
[SIMON - code does not run - "Error in get_downloads(schrader.data):could not find function "get_downloads"".  I assume this is because the Neotoma2 get_downloads function has not yet been made available]
```{r get_singleSitebyName}
schrader <- get_sites(sitename = 'Schrader%') # % can be used to approximate sitenames
schrader  # Look in output to find datasetid for Schrader (27408)
schrader.data <- get_datasets(27408)
schrader.dl <- get_downloads(schrader.data)
schrader.samp <- samples(schrader.dl)
```

#### Count data for one dataset ID

Example of how to get information for a single datasetid using get_dataset using `samples()` to return counts.

```{R ExampleSingleSite }
mirror_data <- get_dataset(39813) # Dataset ID for a diatom surface sample from Mirror Lake; newer dataset/ site ids do not seem to work

mirror_samp <- get_downloads(mirror_data) %>%
  samples()
```

##### Count data using site name or datasetid - if taxon x included
Following code uses Neotoma R - Upper Wallface Pond example
The last few lines of code seem to have the purpose of getting count data only if a particular taxon is present - but it does not seem to work.

```{R getCountsSingleSiteUWallfaceEXAMPLE}
uwallface <-get_sites('Upper Wallface%') # find datasetid for UW
uwall <- get_datasets(32252) # diatom datasetid for UWall is 32252
uwallface.dl <- get_downloads(uwall)
uwallface.samp <- samples(uwallface.dl)
```


##### Count data for a single site name or datasetid

Upper Wallface Pond examples
```{R Upper Wallface EX}
library(neotoma2)
### Find Neotoma Site Id for Upper Wallface Pond 
uwallface <- get_sites(sitename = "Upper Wallface%")
### Find datasets and IDs associated with U Wallface Pond
uwallface.dl<- get_datasets(uwallface)  # Review datasets available note ID of the one of interest
### download data for diatom core dataset (Dataset 32252)
uwallface.dw <- get_downloads(32252)
### Get diatom counts for dataset 32252
uwallface.samp <- samples(uwallface.dw)
```
### Count data for sites in gpids and specific datasettypes
SIMON ASSISTED EXAMPLE [Example provided by Simon Goring - date]
Creates file with diatom counts for surface samples 
```{R datasetByTaxa from Simon}
library(dplyr)
library(purrr)
# Get diatom datasets (Diatom and Diatom Surface Sample data types) for all sites in the northeast US, defined as a set of states.
gpids <- c(8412, 7990, 7934, 7326, 7923, 8981, 6442, 7368)
# gpids for all New England states plus NY and NJ
ne_diatom <- map(gpids, function(x)
  neotoma:::get_dataset(gpid = x,datasettype="diatom")) %>%
  neotoma::bind()   
ne_diatomSS <- map(gpids, function(x)
  neotoma:::get_dataset(gpid = x,datasettype="diatom surface sample"))%>% 
  neotoma::bind() 
ne_dataset<-bind(ne_diatom,ne_diatomSS)
# [Simon, there is an error in one of the following lines of code]
if('DiatomFil' %in% list.files()) {
  Diatom_dl <- readRDS('DiatomFil') 
} else {
  allDatasets <- list()
  for(i in length(allDatasets):length(ne_dataset)) {
    allDatasets[[i]] <- get_download(ne_dataset[[i]])
  }
  ne_download <- neotoma::bind(allDatasets)
  saveRDS(ne_download, 'DiatomFil')
}
# Create diatom count files
wideCounts <- compile_downloads(Diatom_dl) 
minRow <- wideCounts %>% 
  group_by(dataset) %>% 
  summarise(min(depth)) 
```

##### Count data for all "diatom" and "diatom surface sample" datasettypes
Get data from from all datasets of "diatom" and "diatom surface sample" dataset types. Code should eventually be modified to include "diatom top=bottom" datasettype.  That datasettype is not currently recognized by Neotoma R packages.
Example of getting surface sample through filtering and using multiple data set types to get metadata
Note: similar code for getting output of water chemistry data is below
```{R datasetByTaxa with DBT}
## Simon Edited this:
diatomss<- get_datasets(datasettype = 'diatom surface sample')
diatom<- get_datasets(datasettype = 'diatom')
#SIMON - "diatom top-bottom" needs to be added as a choice for datasettype.

#Sapplyy function is used to return datasetids for diatom and diatom surface sample data types.  The function reads the metadata selects datasetids, and puts them in "suset objects".  The query is limited to the first 20 lakes. 
subsetss <- sapply(diatomss, 
                   function(x) {
                     x$dataset.meta$dataset.id 
                   })[1:20]
subset <- sapply(diatom, 
                 function(x) {
                   x$dataset.meta$dataset.id
                 })[1:20]
if('filename' %in% list.files()) {
  diat_dl <- readRDS('filename') # It returns the file that is used 
} else {
  diat_dl <- get_download(subset)
  diat_dlss <- get_download(subsetss)
  diat_dl <- setdiff(diat_dl, diat_dlss)
  saveRDS(diat_dl, 'filename') # It saves this as a download object of the download with file
}
wideCounts <- compile_downloads(diat_dl) 
#neotoma function converts multiple downloads into a single object 
minRow <- wideCounts %>% 
  group_by(dataset) %>% #dataset was returned from the table 
  summarise(min(depth)) #dplyr function to create a dataframe 
#Overall count data is retrieved from this portion 
wideCounts %>% filter(!is.na(Cornus)) # SIMON this line does not work 
#dplyr function used to subset a data frame
diat_taxa <- taxa(diat_dl)# The function taxa returns a data frame of taxa for an (object)
```


Example of returning taxa 

```{R ListTaxaExample}
#searching by taxa not possible in neotoma2 (?)
get_datasets(taxonids = )
get_datasets(datasettype = 'Diatom',taxonname = 'x')
```

Get count data for a single core (neotoma1 example)

```{R Counts single dataset UWall EX}
### Find Neotoma Site Id for Upper Wallface Pond 
uwallface <- get_sites(sitename = "Upper Wallface%")
### Find datasets and IDs associated with U Wallface Pond
uwallface.dl<- get_datasets(uwallface)  # Review datasets available; note ID of the one of interest
### download data for diatom core dataset (Dataset 32252)
uwallface.dw <- get_downloads(32252)
### Get diatom counts for dataset 32252
uwallface.samp <- samples(uwallface.dw)
```
#### Site information and environmental data
Needs to be written

#### Water chemistry
```{R WCHM Out}
uwallf <- get_sites(sitename="%Wallface%")
uwallf_wc <- get_datasets(uwallf,datasettype = 'water chemistry') # example
```

#### `chronologies()` Function
The `chronologies()` function obtains chronology data from a record(s) and returns results as a list. 

**Usage:** `chronologies(x)`

**argument:** (x) is a `sites object` (that contains chronologies) obtained from `get_downloads`.

**Values returned:** A list with the variables chronologyid, notes, contact, agemodel, ageboundolder, ageboundyounger, isdefault (?), dateprepared, modelagetype, chronologyname, and chroncontrols.

```{R chronologies, eval=FALSE}
uwallf_dl <- get_sites(sitename = 'Upper Wallface%') %>%
  get_datasets() %>%
  get_downloads()
 
uwallf_chron <- chronologies(uwallf_dl)
```

#### Other types of data 
- LOI, geochemistry
```{R geochem}
get_datasets(datasettype="geochemistry")
```
Example
```{R ETC EX}
get_datasets(datasettype="loss-on-ignition")
```


Issues:  Put metadata in the same files as the data matrices?  How to combine related files (e.g., diatom counts and chronology)? [When is it best to convert diatom counts to percentages?]

### Input files for data analysis programs
Some programs have no standard input format
#### Bacon, ### Vegan, ### Rioja, ### Tidypaleo
#### Others 



## Modify and analyze data 
- and other miscellaneous tasks
### Add to or modify data in Neotoma
Code being writtent for Neotoma 2

### Add aggregate datasets to Neotoma

### Transform raw diatom counts to percentages and other forms
Use the tran function in the Analogue package to transform raw data.  See analogue package documentation for a list o possible transformations.

``` {R TransformCountsToPct}
#below code works for neotoma1
library ("analogue")
uwallf_dtm_pct <- analogue::tran(x = uwallf_download [[1]]$counts, method = "percent") # Transforms raw counts to percents; uwallf_download is a get_ download file with diatom counts for Upper Wallface Pond
```

``` {R TransformCountsToPct 2}
#manual, no additional package needed
samples(uwallf_download) %>%
pct_abun <- group_by(sampleid) %>%
    mutate(count_pct = count*100 / sum(count))
```

Note: Basic code for above chunk is from S. Goring, J. Williams and E. Grimm Neotoma R Workshop documentation

### Create new ecological groups and assign taxa to the groups
```{R set taxa}
set_taxa('x') # No code yet, but planned
# Example
```

### Add to a table that explicitly matches water chemistry and diatom datasets
This is especially important for cases where sites were sampled for water chemistry as part of more than one study.  As of Oct 2021, corresponding water chemistry and diatom counts are stored in an Excel file.

### Extract a list of publications for a set of datasets
Returns publicationid, citation, and DOI for pub if there is one
```{R get pub out}
# get_publications(datasetid= x)  # Basic format
uwall_pubs <- get_publications(datasetid=32252) # Example - U Wallface P datasetid
uwall_pubs
```

### Extract a list of DOI’s for a set of datasets
```{R DOIS}
#datasetid 32252 doesn't have associated DOIs, choose another?
as.data.frame(uwall_pubs)$doi
```

### Add standard sets of diatom taxa names that could be used to convert multiple synonyms to another set of names
A complicated task - synonym lists need to be prepared and code written
```{R convert taxa}
#code taken from EPD binder
#combines all the taxa with the "Melosira" genus to one group, chosen as simply the genus name "Melosira"

mutate(variablename = replace(variablename,stringr::str_detect(variablename, "Melosira.*"), "Melosira"))
```

## Make plots and maps 
(e.g., stratigraphic diagrams, maps of taxa distributions)

### `plotLeaflet()` function

The `neotoma2` package has a function to map a `sites object`. Use it, for example, to create a map showing all sites that are located within a specific geographic region, are part of an aggregate dataset, etc. 

```{R siteSpatialSearch1}
northeastUS <- rbind(c(-80,40), c(-80,48),
                     c(-65,48), c(-65,40), c(-80,40))
neus <- sf::st_polygon(list(northeastUS))

#As the default, `get_datasets()` only retrieves the first 25 results. Use the argument `limit =" to change this.
neUSsites <- get_datasets(loc = neus, limit = 199)

my_map <- plotLeaflet(neUSsites)

my_map
``` 

The map returned by `plotLeaflet()` is a `leaflet` object. Leaflet items (as defined in the `leaflet` package) can then be added to the map as desired.

```{R siteSpatialSearch2}
library(leaflet)
my_map %>% addCircleMarkers(lng =-80, lat = 45, color = "red") #adds a red marker
``` 



### Map of distribution of a diatom taxon
Option 1 - basic map showing occurrence of a taxon
Option 2 - basic map showing occurrence of multiple taxa
Option 3 - same as option 1, but with symbol size proportional to % abundance
Option 4 - same as option 2, but with symbol size proportional to % abundance

### Plot of the abundance of diatom taxa in a single core
single or multiple taxa - vs depth or time
### Plot of abundance of a single taxon in multiple cores
single or multiple taxa from multiple sites - vs depth or time
### Plot of abundance of a taxon vs an environmental characteristic 
(e.g., water chemistry variable)

## Get taxonomic data from DiatomBase

DiatomBase is a subset of WORMS - World Register of Marine Species
https://www.diatombase.org/index.php
http://www.marinespecies.org/index.php

`worrms` is an R client for the World Register of Marine Species (<http://www.marinespecies.org/>).  See the taxize book (<https://taxize.dev>) for taxonomically focused work in this and similar packages.   

DiatomBase API's are at: https://www.diatombase.org/aphia.php?p=rest&__route__/.  They are the same as those for WORMS, but functions are restricted to returning records of diatom taxa only.

The following basic R code and text was copied on 29 Oct 2021 by DF Charles from: https://github.com/ropensci/worrms/blob/master/vignettes/worrms.Rmd

Install stable version from CRAN
```r
install.packages("worrms")
```
Development version from GitHub
```r
install.packages("remotes")
remotes::install_github("ropensci/worrms")
```

```r
library("worrms")
```
Notes (dfc): 
The unique identifier for a taxon name in WORMS and DiatomBase is called      "AphiaID"
The code below uses three functions:
* wm_records_name - gets primary data on one taxon
- wm_records_names - gets primary data on two or more taxa
- wm_sources - gets source information, including reference citation  for the original description

The following website provides a useful way to convert a column  of taxa names to a comma-separated list.  Change the preferences to get a list  with quotes around each name:  https://convert.town/column-to-comma-separated-list

Purpose of code chunks in this section - in the order presented
1.  Get AphiaID and taxonomic information by one taxon name
2.  Get AphiaIDs and taxonomic information by multiple taxa names
3.  Create an object that includes the AphiaID's only from results of 2, above
4.  Get source data, including original description, by AphiaIDs (contained in object created in number 3, above.
5.  Combine results of 2 and 4, above, to get information necessary to populate all fields in the Neotoma taxa list.

To get an Excel file with data for entering names in the Neotoma taxa list:
1. Start with a column of taxa names, convert it to a comma-delimited list
2. Convert the column of names to a comma delimited file with each name in quotes (e.g., by using convert.town)
3. Run code chunks (described above) in the following order: 2, 3, 4, 5.

### By single taxon name
```{r DiatomBase1Taxon}
library("worrms")
dtmBase_taxon <- wm_records_name(name = 'Neofragilaria acidobiontica', fuzzy = TRUE, marine_only = FALSE) 
# Set fuzzy = TRUE (default) to include varieties, etc.
# Set fuzzy = FALSE to get specified name only - must be exact matches
# Set marine-only as FALSE; default value is TRUE
view (dtmBase_taxon)
```
### By multiple taxa names

```{r DiatomBaseTaxa}
library("worrms")
dtmBase_taxa <- wm_records_names(name = c("Cyclotella chochtawatcheana","Cyclotella meneghiniana","Cyclotella stelligera","Melosira cf. arenti","Aulacoseira sp.","Cyclostephanos sp.","Centric spp.","Chaetoceros sp. ","Pennate Diatoms","Achnanthes brevipes","Achnanthes delicatula","Achnanthes cf. engel ","Achnanthes cf. grana","Achnanthes lanceolata","Achnanthes minutissima","Achnanthes spp.","Amphora coffeiformis","Amphora veneta","Amphora spp.","Caloneis spp.","Cocconeis placentula and varieties","Cocconeis other spp.","Cymbella spp.","Denticula subtilis","Diploneis spp.","Entomoneis spp.","Epithemia spp.","Fragilaria brevistriata","Fragilaria capucina var. vaucheriae","Fragilaria construens and varieties","Fragilaria leptostauron","Fragilaria pinnata","Fragilaria subsalina","Fragilaria zelleri","Fragilaria spp.","Gomphonema","Gyrosigma"), fuzzy = FALSE, marine_only = FALSE)
# Set fuzzy = TRUE (default) to include varieties, etc.
# Set fuzzy = FALSE to get specified name only - must be exact match
# Set marine-only as FALSE; default value is TRUE
dtmBase_out <- bind_rows(dtmBase_taxa)
view(dtmBase_out)
WriteXLS(dtmBase_out, ExcelFileName = "DtmBase_taxa_out.xlsx")
```

### Get AphiaIDs

```{r DiatomBaseAphiaIDs}
# Create an object with AphiaIDs using output from the code chunk above
db_out_Aphiaid <- dtmBase_out$AphiaID
view (db_out_Aphiaid)
```

### Create output file for Neotoma

```{r DiatomBaseSources}
# Create an object combining primary information and reference citations  for taxa with specified AphiaIDs
# Get source information for the set of AphiaIDs created in the  code chunk above.
dtmBase_source <- wm_sources_(id = db_out_Aphiaid)
# Returns sources for several variables, one of which is specified as "original description" in the "use" column
# Create an object with source for "original description" only
dtm_Base_origDes <- filter(dtmBase_source, use == "original description")
# Limit variables / columns to those most relevant
dtm_Base_origDes1 <- select(dtm_Base_origDes, id ,reference, page, doi)
# Change "id" column from character to integer type
dtm_Base_origDes1$id <- as.integer(dtm_Base_origDes1$id)
# Combine main taxa table (dtmBase_out) and source table (dtm_Base_origDes1)
dtm_Base_comb <- left_join(dtmBase_out, dtm_Base_origDes1, by = c("AphiaID" = "id"))
# Create table with variables needed for adding names to Neotoma taxa list
dtmBase_neo <- select (dtm_Base_comb, "AphiaID", "scientificname", "authority","status", "unacceptreason", "reference", "page", "doi", "citation", "rank", "genus", "family", "valid_AphiaID", "valid_name", "valid_authority")
view (dtmBase_neo)
# Create an Excel file
WriteXLS(dtmBase_neo, ExcelFileName = "DtmBase_addNeotoma.xlsx")
```
